File: julia/benchmark/bench_mps.jl
# Usage: julia --project=../julia bench_mps.jl <data.json> [N=10]

include("../src/Simplex.jl")
using .Simplex
using JSON3

function main()
    length(ARGS) >= 1 || error("Usage: bench_mps.jl <data.json> [N=10]")
    data_path = ARGS[1]
    N = length(ARGS) >= 2 ? parse(Int, ARGS[2]) : 10

    data = JSON3.read(read(data_path, String))

    A = reduce(hcat, [Float64.(collect(row)) for row in data["A"]])'  # m×n
    b = Float64.(collect(data["b"]))
    c = Float64.(collect(data["c"]))

    lp = LinearProgram(A, b, c)

    # Warmup solves (Julia JIT)
    solve(lp)
    solve(lp)

    times_ms = Vector{Float64}(undef, N)
    local result
    for i in 1:N
        t0 = time_ns()
        result = solve(lp)
        times_ms[i] = (time_ns() - t0) / 1e6
    end

    z = result.status == OPTIMAL ? result.z : nothing
    iters = result.iterations

    times_str = join(times_ms, ", ")
    if isnothing(z)
        print("{\"z\": null, \"iterations\": $iters, \"times_ms\": [$times_str]}")
    else
        print("{\"z\": $z, \"iterations\": $iters, \"times_ms\": [$times_str]}")
    end
end

main()

File: python/benchmark/bench_mps.py
# Usage: python bench_mps.py <data.json> [N=10]

from __future__ import annotations

import json
import sys
import time

import numpy as np

from simplex import LinearProgram, SolveStatus, solve


def main() -> None:
    if len(sys.argv) < 2:
        print("Usage: bench_mps.py <data.json> [N=10]", file=sys.stderr)
        sys.exit(1)

    data_path = sys.argv[1]
    N = int(sys.argv[2]) if len(sys.argv) >= 3 else 10

    with open(data_path) as f:
        data = json.load(f)

    A = np.array(data["A"], dtype=float)
    b = np.array(data["b"], dtype=float)
    c = np.array(data["c"], dtype=float)

    lp = LinearProgram(A, b, c)

    times_ms: list[float] = []
    result = None
    for _ in range(N):
        t0 = time.perf_counter()
        result = solve(lp)
        times_ms.append((time.perf_counter() - t0) * 1e3)

    z = result.z if result is not None and result.status == SolveStatus.OPTIMAL else None
    iters = result.iterations if result is not None else 0

    print(json.dumps({"z": z, "iterations": iters, "times_ms": times_ms}))


if __name__ == "__main__":
    main()
